File: python-soa/inventory-service/app/infrastructure/messaging/events.py
"""
Inventory Events
==================
Factory methods for creating inventory-related domain events.

Events Published:
- stock.reserved: Stock successfully reserved (saga response - happy path)
- stock.insufficient: Cannot reserve stock (saga response - compensation trigger)
- stock.released: Stock released after order cancellation
"""

from datetime import datetime
from uuid import UUID


class InventoryEvents:
    """Factory class for creating inventory-related domain events."""

    @staticmethod
    def stock_reserved(order_id: UUID, items: list[dict]) -> dict:
        """
        Create a 'stock.reserved' event payload.

        Published when stock is successfully reserved for all order items.
        The Order Service consumes this to confirm the order.
        """
        return {
            "event_type": "stock.reserved",
            "timestamp": datetime.utcnow().isoformat(),
            "data": {
                "order_id": str(order_id),
                "items": [
                    {
                        "product_id": item["product_id"],
                        "quantity": item["quantity"],
                    }
                    for item in items
                ],
            },
        }

    @staticmethod
    def stock_insufficient(order_id: UUID, reason: str) -> dict:
        """
        Create a 'stock.insufficient' event payload.

        Published when stock cannot be reserved for the order.
        The Order Service consumes this to cancel the order (compensation).
        """
        return {
            "event_type": "stock.insufficient",
            "timestamp": datetime.utcnow().isoformat(),
            "data": {
                "order_id": str(order_id),
                "reason": reason,
            },
        }

    @staticmethod
    def stock_released(order_id: UUID, items: list[dict]) -> dict:
        """
        Create a 'stock.released' event payload.

        Published after releasing reserved stock due to order cancellation.
        Informational event for audit/monitoring purposes.
        """
        return {
            "event_type": "stock.released",
            "timestamp": datetime.utcnow().isoformat(),
            "data": {
                "order_id": str(order_id),
                "items": [
                    {
                        "product_id": item["product_id"],
                        "quantity": item["quantity"],
                    }
                    for item in items
                ],
            },
        }

File: dotnet-soa/InventoryService/Infrastructure/Messaging/Events.cs
namespace InventoryService.Infrastructure.Messaging;

/// <summary>
/// Domain events for the Inventory Service.
/// 
/// Published events:
/// - stock.reserved: Stock successfully reserved for an order
/// - stock.insufficient: Not enough stock available for an order
/// - stock.released: Reserved stock released after order cancellation
/// 
/// Consumed events:
/// - order.created: New order requiring stock reservation
/// - order.cancelled: Order cancelled, release reserved stock
/// </summary>

/// <summary>Published when stock is successfully reserved</summary>
public record StockReservedEvent
{
    public Guid OrderId { get; init; }
    public DateTime Timestamp { get; init; } = DateTime.UtcNow;
}

/// <summary>Published when stock is insufficient for an order</summary>
public record StockInsufficientEvent
{
    public Guid OrderId { get; init; }
    public string Reason { get; init; } = string.Empty;
    public DateTime Timestamp { get; init; } = DateTime.UtcNow;
}

/// <summary>Published when reserved stock is released</summary>
public record StockReleasedEvent
{
    public Guid OrderId { get; init; }
    public DateTime Timestamp { get; init; } = DateTime.UtcNow;
}

/// <summary>Consumed from Order Service when a new order is created</summary>
public record OrderCreatedEvent
{
    public Guid OrderId { get; init; }
    public string CustomerName { get; init; } = string.Empty;
    public string CustomerEmail { get; init; } = string.Empty;
    public List<OrderItemEvent> Items { get; init; } = new();
    public decimal TotalAmount { get; init; }
    public DateTime Timestamp { get; init; }
}

/// <summary>Order item within an order event</summary>
public record OrderItemEvent
{
    public Guid ProductId { get; init; }
    public string ProductName { get; init; } = string.Empty;
    public int Quantity { get; init; }
    public decimal UnitPrice { get; init; }
}

/// <summary>Consumed from Order Service when an order is cancelled</summary>
public record OrderCancelledEvent
{
    public Guid OrderId { get; init; }
    public string Reason { get; init; } = string.Empty;
    public DateTime Timestamp { get; init; }
}
